Python/Product/TestAdapter.Executor/Pytest/TestDiscovererPytest.cs

```diff
@@ -25,52 +41,94 @@ internal class TestDiscovererPytest : IPythonTestDiscoverer {
+                json = ExecuteProcess(
                     _settings.InterpreterPath,
+                    env,
                     arguments,
                     _settings.WorkingDirectory,
-                    env,
-                    visible: false,
-                    new StreamRedirector(writer)
-                )) {
+                    _settings.PathEnv,
+                    _settings.DiscoveryWaitTimeInSeconds
+                    );
+            } catch (TimeoutException) {
+                Error(Strings.PythonTestDiscovererTimeoutErrorMessage);
+                return;
+            }
 
-                    DebugInfo("cd " + _settings.WorkingDirectory);
-                    DebugInfo("set " + _settings.PathEnv + "=" + env[_settings.PathEnv]);
-                    DebugInfo(proc.Arguments);
+            List<PytestDiscoveryResults> results = null;
+            try {
+                results = JsonConvert.DeserializeObject<List<PytestDiscoveryResults>>(json);
+            } catch (InvalidOperationException ex) {
+                Error("Failed to parse: {0}".FormatInvariant(ex.Message));
+                Error(json);
+            } catch (JsonException ex) {
+                Error("Failed to parse: {0}".FormatInvariant(ex.Message));
+                Error(json);
+            }
 
-                    // If there's an error in the launcher script,
-                    // it will terminate without connecting back.
-                    WaitHandle.WaitAny(new WaitHandle[] { proc.WaitHandle });
+            CreateVsTests(results, discoverySink);
+        }
 
-                    outputStream.Flush();
-                    outputStream.Seek(0, SeekOrigin.Begin);
-                    var json = new StreamReader(outputStream).ReadToEnd();
 
+        private void CreateVsTests(IEnumerable<PytestDiscoveryResults> discoveryResults, ITestCaseDiscoverySink discoverySink) {
+            foreach (PytestDiscoveryResults result in discoveryResults ?? Enumerable.Empty<PytestDiscoveryResults>()) {
+                var parentMap = result.Parents.ToDictionary(p => p.Id, p => p);
+                foreach (PytestTest test in result.Tests) {
                     try {
-                        discoveryResults = JsonConvert.DeserializeObject<List<PytestDiscoveryResults>>(json);
-                    } catch (InvalidOperationException ex) {
-                        Error("Failed to parse: {0}".FormatInvariant(ex.Message));
-                        Error(json);
-                    } catch (JsonException ex) {
-                        Error("Failed to parse: {0}".FormatInvariant(ex.Message));
-                        Error(json);
+                        TestCase tc = test.ToVsTestCase(_settings.ProjectHome, parentMap);
+                        DebugInfo($"{tc.DisplayName} Source:{tc.Source} Line:{tc.LineNumber}");
+                        discoverySink?.SendTestCase(tc);
+                    } catch (Exception ex) {
+                        Error(ex.Message);
```

Comment:
this code was moved from previous   PyTestDiscoveryReader.ParseDiscovery(discoveryResults, discoverySink, _settings, logger);

but now that PytestDiscoveryResults is no longer a struct we need to check for null with ??